src/usr/share/netcheck/networkmanagerhelper.py

```diff
@@ -0,0 +1,304 @@
+
+                if gateway_address in address_network:
+                    ip_address = address_data['address']
+
+            if ip_address is None:
+                self.logger.warning(
+                    'No IP addresses for connection %s associated with gateway %s.',
+                    connection_id, gateway_address)
+
+        return ip_address
+
+    def connection_is_activated(self, connection_id):
+        """Check whether the connection with the given connection ID is activated.
+
+        connection_id: The displayed name of the connection in NetworkManager.
+        Returns True if the connection is activated, False otherwise.
+        """
+
+        connection_is_activated = False
+
+        connection_state = self._get_connection_state(connection_id)
+
+        if connection_state is NM_CONNECTION_ACTIVATED:
+            connection_is_activated = True
+
+        return connection_is_activated
+
+    def _build_connection_id_dict(self):
+        """Assemble a dictionary of connection objects, indexed by the connection's ID in
+        NetworkManager.
+
+        Returns a dict, with keys being connection IDs and values being
+          NetworkManager.Connection objects.
+        """
+        connection_id_to_connection_dict = {}
+
+        all_connections = NetworkManager.Settings.ListConnections()
+        self._run_proxy_call(all_connections)
+
+        for connection in all_connections:
+            connection_id = connection.GetSettings()['connection']['id']
+            connection_id_to_connection_dict[connection_id] = connection
+
+        return connection_id_to_connection_dict
+
+    def _create_device_objects(self, wired_interface_name, wireless_interface_name):
+        """Store references to the device objects that are used frequently.
+
+        wired_interface_name: The name of the wired network interface, ex: eth0, enp0s1.
+        wireless_interface_name: The name of the wireless network interface, ex: wlan0,
+          wlp0s1.
+        Returns a tuple containing the NetworkManager wired device object and the
+          NetworkManager wireless device object.
+        """
+        wired_device = None
+        wireless_device = None
+
+        for device in NetworkManager.NetworkManager.GetDevices():
+            if device.Interface == wired_interface_name:
+                wired_device = device
+
+            if device.Interface == wireless_interface_name:
+                wireless_device = device
+
+        if wired_device is None:
+            raise DeviceNotFoundException('Configured wired device %s was not found.' %
+                                          wired_interface_name)
+
+        if wireless_device is None:
+            raise DeviceNotFoundException('Configured wireless device %s was not found.' %
+                                          wireless_interface_name)
+
+        return (wired_device, wireless_device)
+
+    def _get_device_for_connection(self, connection_id):
```

Comment:
I'm not sure if caching the device is a good idea. I'm having a couple problems with it.

I tested this with a USB ethernet adapter. If the ethernet adapter is not present when the daemon starts, wired_device ends up being None. Also, if I remove the ethernet adapter and reattach it, the device ends up with a different dbus object path and is no longer accessible through the wired_device variable.

You could argue that the user should just not remove the ethernet adapter, however, if we did this, we could no longer claim this is a super robust piece of software, could we?